drivers/sensor/bq34110/bq34110.c

```diff
@@ -0,0 +1,381 @@
+		{ FULL_CHARGE_CAPACITY, &priv->full_charge_capacity },
+		{ REMAINING_CAPACITY, &priv->remaining_charge_capacity },
+	};
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+	for (size_t i = 0; i < ARRAY_SIZE(regs); i++) {
+		int rc;
+
+		rc = bq34110_cmd_reg_read(dev, regs[i].reg_addr,
+					  regs[i].dest);
+		if (rc != 0) {
+			LOG_ERR("Failed to read channel %d", chan);
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+static int bq34110_ctrl_reg_read(const struct device *dev,
+				 uint16_t subcommand,
+				 uint16_t *val, uint8_t read_addr)
+{
+	int ret;
+	uint8_t write_data[3], read_data[2];
+	struct bq34110_data *priv = dev->data;
+	const struct bq34110_config *cfg = dev->config;
+
+	write_data[0] = MANUFACTURER_ACCESS_CONTROL;
+	write_data[1] = (uint8_t)(subcommand & 0x00FF);
+	write_data[2] = (uint8_t)((subcommand >> 8) & 0x00FF);
+	ret = i2c_write(priv->i2c, write_data, sizeof(write_data),
+			cfg->i2c_addr);
+	if (ret < 0) {
+		LOG_ERR("Failed to write");
+		return ret;
+	}
+
+	k_msleep(BQ34110_DELAY);
+
+	ret = i2c_write_read(priv->i2c, cfg->i2c_addr, &read_addr,
+			     sizeof(read_addr), &read_data, sizeof(read_data));
+	if (ret < 0) {
+		LOG_ERR("Failed to read write");
+		return ret;
+	}
+
+	*val = ((read_data[1] << 8) | read_data[0]);
+
+	return 0;
+}
+
+static int update_df_parameter(const struct device *dev, uint16_t df_addr,
+			       uint8_t data[], uint8_t len)
+{
+	int ret;
+	uint16_t datasum;
+	uint8_t i, df_addr_data[3],
+		datasum_len[3], df_write_data[len + 1];
+	struct bq34110_data *priv = dev->data;
+	const struct bq34110_config *cfg = dev->config;
+
+	df_addr_data[0] = MANUFACTURER_ACCESS_CONTROL;
+	df_addr_data[1] = (uint8_t)(df_addr & 0x00FF);
+	df_addr_data[2] = (uint8_t)((df_addr >> 8) & 0x00FF);
+
+	ret = i2c_write(priv->i2c, df_addr_data,
+			sizeof(df_addr_data), cfg->i2c_addr);
+	if (ret < 0) {
+		LOG_ERR("Failed to write MAC");
+		return ret;
+	}
+
+	df_write_data[0] = MAC_DATA;
+	for (i = 1; i <= len; i++) {
+		df_write_data[i] = data[i-1];
+	}
+
+	ret = i2c_write(priv->i2c, df_write_data, sizeof(df_write_data),
+			cfg->i2c_addr);
+	if (ret < 0) {
+		LOG_ERR("Failed to write MAC");
+		return ret;
+	}
+
+	datasum = df_addr_data[1] + df_addr_data[2];
+	for (i = 1; i <= len; i++) {
+		datasum += df_write_data[i];
+	}
+
+	datasum_len[0] = MAC_DATA_SUM;
+	datasum_len[1] = (uint8_t)~datasum;
+	datasum_len[2] = 4 + len;
+	ret = i2c_write(priv->i2c, datasum_len, sizeof(datasum_len),
+			cfg->i2c_addr);
+	if (ret < 0) {
+		LOG_ERR("Failed to write MACDatasum");
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief initialise the fuel gauge
+ *
+ * @return 0 for success
+ */
+static int bq34110_gauge_init(const struct device *dev)
+{
+	int ret;
+	uint8_t writedata[2];
+	int16_t design_capacity;
+	struct bq34110_data *priv = dev->data;
+	const struct bq34110_config *cfg = dev->config;
+	uint16_t design_voltage, device_type, taper_current, taper_voltage;
+
+	priv->i2c = device_get_binding(cfg->bus_name);
```

Comment:
Resolve the device at build time instead of runtime by using `I2C_DT_SPEC_INST_GET` in `BQ34110_INIT`